### PR TITLE
Implementing names for characters

### PR DESCRIPTION
**Pull request by agent-q1**
This is a basic implementation of names on characters.
There are still a bunch of logistic questions to be answered like whether we want to display the type of citizens too in the Tooltips. Would love to hear the community's opinions on this
Further, the names are being taken from a list of ELVEN names prefab in Name Generator. We could definitely add a prefab with Citizen names. Also the shopkeeper's name is rendered as a shopkeeper itself (again something I'd like to hear your opinions on).
And as always, a review is much appreciated :).

Added https://github.com/Terasology/NameGenerator/pull/23 to Hopefully use Country names instead of ELVEN names, However sometimes, Names that do not belong to the list pop up in the citizen Tooltips.
Might need some testing and bug finding.

### CODE REVIEW

**Comment by AndyTechGuy** [comment]
Pull request works exactly as stated, names are generated for each character and displayed on the tooltip, very well done! Some ideas that I have for some of the logistic questions you proposed:

1. LIke you already suggested, a prefab with more fitting Citizen names would be very nice; the elven names don't fit very well and aren't really memorable.
2. I think that the shopkeeper should also have a name, it would be really cool if the word "Shopkeeper" could be used as a kind of prefix, like "Shopkeeper Jones", for example.
3. I think that the type of citizen can be ignored in the tooltip; the type is implied by the character's appearance anyways

**Review by mayant15 — CHANGES_REQUESTED** [review]
I think you've had a little git problem here. This PR now also contains changes from your character overlay PR.

File: src/main/java/org/terasology/metalrenegades/ai/system/CitizenSpawnSystem.java [hunk 1]
```diff
@@ -39,10 +40,16 @@
...
+import org.terasology.metalrenegades.minimap.events.AddCharacterToOverlayEvent;
+import org.terasology.namegenerator.creature.CreatureAssetTheme;
+import org.terasology.namegenerator.creature.CreatureNameProvider;
 import org.terasology.registry.In;
 
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Random;
+
+import static org.terasology.namegenerator.creature.CreatureAssetTheme.COUNTRY;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/metalrenegades/ai/system/CitizenSpawnSystem.java]: As I've requested to change the name list prefix to "Old West" this (and probably other spots) need changes now... 🙄 I'm eager to merge https://github.com/Terasology/NameGenerator/pull/23, though (after renames) 👍 

```suggestion
import static org.terasology.namegenerator.creature.CreatureAssetTheme.OLD_WEST;
```

File: src/main/java/org/terasology/metalrenegades/ai/system/CitizenSpawnSystem.java [hunk 2]
```diff
@@ -124,6 +134,12 @@ private EntityRef spawnCitizen(EntityRef homeEntity) {
 
         EntityRef entityRef = entityBuilder.build();
 
+        if(entityRef.hasComponent(DisplayNameComponent.class)) {
+            DisplayNameComponent displayNameComponent = entityRef.getComponent(DisplayNameComponent.class);
+            CreatureNameProvider creatureNameProvider = new CreatureNameProvider(random.nextLong(), CreatureAssetTheme.COUNTRY);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/metalrenegades/ai/system/CitizenSpawnSystem.java]:
```suggestion
            CreatureNameProvider creatureNameProvider = new CreatureNameProvider(random.nextLong(), CreatureAssetTheme.OLD_WEST);
```

File: src/main/java/org/terasology/metalrenegades/economy/systems/MarketCitizenSpawnSystem.java [hunk 3]
```diff
@@ -65,9 +72,15 @@ public void onMarketPlaceSpawn(BuildingEntitySpawnedEvent event, EntityRef entit
                 Rect2i rect2i = dynParcel.shape;
                 Vector3f spawnPosition = new Vector3f(rect2i.minX() + rect2i.sizeX() / 2, dynParcel.getHeight() + 1, rect2i.minY() + rect2i.sizeY() / 2);
                 EntityRef trader = entityManager.create(traderGooeyOptional.get(), spawnPosition);
+                trader.send(new AddCharacterToOverlayEvent());
                 SettlementRefComponent settlementRefComponent = entityRef.getComponent(SettlementRefComponent.class);
                 trader.addComponent(settlementRefComponent);
                 MarketComponent marketComponent = settlementRefComponent.settlement.getComponent(MarketComponent.class);
+                if(entityRef.hasComponent(DisplayNameComponent.class)) {
+                    DisplayNameComponent displayNameComponent = entityRef.getComponent(DisplayNameComponent.class);
+                    CreatureNameProvider creatureNameProvider = new CreatureNameProvider(random.nextLong(), CreatureAssetTheme.COUNTRY);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/metalrenegades/economy/systems/MarketCitizenSpawnSystem.java]:
```suggestion
                    CreatureNameProvider creatureNameProvider = new CreatureNameProvider(random.nextLong(), CreatureAssetTheme.OLD_WEST);
```

**Review by skaldarnar — CHANGES_REQUESTED** [review]
As already noted, there are unrelated changes in this PR (character overlay). Please rebase or re-apply only the changes related to adding the names.

File: src/main/java/org/terasology/metalrenegades/ai/system/CitizenSpawnSystem.java [hunk 4]
```diff
@@ -124,6 +132,12 @@ private EntityRef spawnCitizen(EntityRef homeEntity) {
 
         EntityRef entityRef = entityBuilder.build();
 
+        if(entityRef.hasComponent(DisplayNameComponent.class)) {
+            DisplayNameComponent displayNameComponent = entityRef.getComponent(DisplayNameComponent.class);
+            CreatureNameProvider creatureNameProvider = new CreatureNameProvider(random.nextLong(), OLD_WEST);
+            displayNameComponent.name = creatureNameProvider.generateName();
```
Comment (skaldarnar) [inline src/main/java/org/terasology/metalrenegades/ai/system/CitizenSpawnSystem.java:138]: Do you also need to save the component again?

File: src/main/java/org/terasology/metalrenegades/economy/systems/MarketCitizenSpawnSystem.java [hunk 5]
```diff
@@ -65,9 +71,15 @@ public void onMarketPlaceSpawn(BuildingEntitySpawnedEvent event, EntityRef entit
...
                 Vector3f spawnPosition = new Vector3f(rect2i.minX() + rect2i.sizeX() / 2, dynParcel.getHeight() + 1, rect2i.minY() + rect2i.sizeY() / 2);
                 EntityRef trader = entityManager.create(traderGooeyOptional.get(), spawnPosition);
+
                 SettlementRefComponent settlementRefComponent = entityRef.getComponent(SettlementRefComponent.class);
                 trader.addComponent(settlementRefComponent);
                 MarketComponent marketComponent = settlementRefComponent.settlement.getComponent(MarketComponent.class);
+                if(entityRef.hasComponent(DisplayNameComponent.class)) {
+                    DisplayNameComponent displayNameComponent = entityRef.getComponent(DisplayNameComponent.class);
+                    CreatureNameProvider creatureNameProvider = new CreatureNameProvider(random.nextLong(), CreatureAssetTheme.OLD_WEST);
+                    displayNameComponent.name = "shopkeeper " + creatureNameProvider.generateName();
```
Comment (skaldarnar) [inline src/main/java/org/terasology/metalrenegades/economy/systems/MarketCitizenSpawnSystem.java:81]: Same here - save the component after modifying it?

File: src/main/java/org/terasology/metalrenegades/ai/system/CitizenSpawnSystem.java (hunk 4 above)
Comment (agent-q1) [reply src/main/java/org/terasology/metalrenegades/ai/system/CitizenSpawnSystem.java:138]: Yea, will add that. However, I feel like it doesn't really make a difference here.

**Comment by AndyTechGuy** [comment]
Closed, this feature is implemented as part of #74.

**Pull request closed without merging**